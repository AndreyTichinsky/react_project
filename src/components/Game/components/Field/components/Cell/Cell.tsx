import React, { FC, PureComponent } from "react";
import type { CellProps } from "types/field";
import { CellItem } from "./CellItem";

export class Cell extends React.PureComponent<CellProps> {
  handleClick = () => {
    this.props.onClick(this.props.x, this.props.y);
  };
  render() {
    return (
      <CellItem
        className={`cell cell_${this.props.y}_${this.props.x}`}
        isAlive={this.props.isAlive}
        x={this.props.x}
        y={this.props.y}
        cellSize={this.props.cellSize}
        onClick={this.handleClick}
        animationSpeed={this.props.animationSpeed}
      />
    );
  }
}
