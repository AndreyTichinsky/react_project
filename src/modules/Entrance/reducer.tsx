import { createSlice, PayloadAction } from "@reduxjs/toolkit";

export type EntranceState = {
  username: string;
};

export const initialState: EntranceState = {
  username: "Guest",
};

export const entranceSlice = createSlice({
  name: "ENTRANCE",
  initialState,
  reducers: {
    setUsername: (state, action: PayloadAction<any>) => {
      return {
        ...state,
        username: action.payload,
      };
    },
    logout: (state) => {
      return {
        ...state,
        username: "Guest",
      };
    },
  },
});

export const { reducer, actions } = entranceSlice;
