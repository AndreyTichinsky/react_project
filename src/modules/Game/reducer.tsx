import { createSlice, PayloadAction } from "@reduxjs/toolkit";
import { FieldData } from "types/game";
import * as helper from "@/modules/Game/GameHelper";

export type GameState = {
  xSize: number;
  ySize: number;
  updateSpeed: string;
  initialPercent: number;
  gameInProgress: boolean;
  cellSize: number;
  fieldState: FieldData;
};

export const initialState: GameState = {
  initialPercent: 0,
  xSize: 5,
  ySize: 5,
  cellSize: 10,
  updateSpeed: "slow",
  gameInProgress: false,
  fieldState: helper.makeMatrix(5, 5, 0),
};

export const gameSlice = createSlice({
  name: "GAME",
  initialState,
  reducers: {
    setCellState: (state, action: PayloadAction<any>) => {
      const { payload } = action;
      const fieldStateCopy = state.fieldState.slice();
      fieldStateCopy[payload] = fieldStateCopy[payload] === 1 ? 0 : 1;
      return {
        ...state,
        fieldState: fieldStateCopy,
      };
    },
    setProgress: (state) => {
      return { ...state };
    },
    stop: (state) => {
      return {
        ...state,
        gameInProgress: false,
      };
    },
    play: (state) => {
      return {
        ...state,
        gameInProgress: true,
      };
    },
    setInitialPercent: (state, action: PayloadAction<any>) => {
      return {
        ...state,
        initialPercent: action.payload,
      };
    },
    setXSize: (state, action: PayloadAction<any>) => {
      return {
        ...state,
        xSize: action.payload,
      };
    },
    setYSize: (state, action: PayloadAction<any>) => {
      return {
        ...state,
        ySize: action.payload,
      };
    },
    setSpeed: (state, action: PayloadAction<any>) => {
      return { ...state, updateSpeed: action.payload };
    },
    setFieldState: (state, action: PayloadAction<any>) => {
      return { ...state, fieldState: action.payload };
    },
    reset: (state) => {
      return { ...state };
    },
    generateRandomField: (state) => {
      return { ...state };
    },
    updateFieldState: (state, action: PayloadAction<any>) => {
      return { ...state, fieldState: action.payload };
    },
    setAnyState: (state, action: PayloadAction<any>) => {
      return { ...state, ...action.payload };
    },
  },
});

export const { reducer, actions } = gameSlice;
