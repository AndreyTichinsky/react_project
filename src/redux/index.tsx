export { reducer } from "./reducer";
export { preloadedState, store, dispatch } from "./store";
export type { State, AppDispatch } from "./store";
