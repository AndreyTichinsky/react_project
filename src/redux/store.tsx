import { configureStore } from "@reduxjs/toolkit";
import createSagaMiddleware from "redux-saga";

import { reducer } from "./reducer";
import { FieldData } from "types/game";
import {
  filledPercentValidation,
  sizeValidation,
  rootSaga,
  helper,
} from "@/modules";

export interface State {
  game: {
    xSize: number;
    ySize: number;
    updateSpeed: string;
    initialPercent: number;
    gameInProgress: boolean;
    cellSize: number;
    fieldState: FieldData;
  };
  entrance: {
    username: string;
  };
}

export const preloadedState: State = {
  game: {
    initialPercent: 0,
    xSize: 5,
    ySize: 5,
    cellSize: 10,
    updateSpeed: "slow",
    gameInProgress: false,
    fieldState: helper.makeEmptyMatrix(5, 5),
  },
  entrance: {
    username: "Guest",
  },
};

const sagaMiddleware = createSagaMiddleware();

export const store = configureStore({
  reducer,
  middleware: [filledPercentValidation, sizeValidation, sagaMiddleware],
  devTools: true,
});

sagaMiddleware.run(rootSaga);

export const { dispatch } = store;
export type AppDispatch = typeof store.dispatch;
