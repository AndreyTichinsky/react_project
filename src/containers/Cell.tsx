import { connect } from "react-redux";
import { Cell } from "@/components/Game/components/Field/components/Cell/Cell";
import { State } from "@/redux/reducers";
import { makeConvertSpeedToNumber, makeGetAlive } from "@/redux/selectors";

const makeMapStateToProps = () => {
  const convertSpeedToNumber = makeConvertSpeedToNumber();
  const getAlive = makeGetAlive();
  const mapStateToProps = (state: State, props: any) => {
    return {
      cellSize: state.cellSize,
      isAlive: getAlive(state, props),
      animationSpeed: convertSpeedToNumber(state),
    };
  };
  return mapStateToProps;
};

export default connect(makeMapStateToProps)(Cell);
