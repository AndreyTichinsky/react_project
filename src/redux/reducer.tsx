import { combineReducers } from "redux";
import { gameSlice } from "@/modules/Game/reducer";
import { entranceSlice } from "@/modules/Entrance/reducer";

export const reducer = combineReducers({
  entrance: entranceSlice.reducer,
  game: gameSlice.reducer,
});
