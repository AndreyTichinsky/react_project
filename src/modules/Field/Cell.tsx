import React, { FC, useCallback } from "react";
import { CellItem } from "./CellItem";
import { connect } from "react-redux";
import { AppDispatch } from "@/redux/store";
import { State } from "@/redux/store";
import { actions } from "@/modules/Game/reducer";
import {
  makeConvertSpeedToNumber,
  makeGetAlive,
} from "@/modules/Game/selectors";

export interface CellProps {
  isAlive: boolean;
  x: number;
  y: number;
  animationSpeed: number;
  idx: number;
}

const makeMapStateToProps = () => {
  const convertSpeedToNumber = makeConvertSpeedToNumber();
  const getAlive = makeGetAlive();
  const mapStateToProps = (state: State, props: any) => {
    return {
      isAlive: !!getAlive(state, props),
      animationSpeed: convertSpeedToNumber(state),
    };
  };
  return mapStateToProps;
};

export const CellComponent: FC<CellProps & { dispatch: AppDispatch }> = (
  props
) => {
  const onClick = useCallback(() => {
    props.dispatch(actions.setCellState(props.idx));
  }, []);
  return (
    <CellItem
      className={`cell cell_${props.y}_${props.x}`}
      isAlive={props.isAlive}
      y={props.y}
      x={props.x}
      onClick={onClick}
      animationSpeed={props.animationSpeed}
    />
  );
};

export const Cell = connect(makeMapStateToProps)(CellComponent);
