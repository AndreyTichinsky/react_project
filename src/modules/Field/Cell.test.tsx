import React from "react";
import renderer from "react-test-renderer";
import { Cell } from "./Cell";
import { Provider } from "react-redux";
import { configureStore } from "@reduxjs/toolkit";
import { preloadedState } from "@/redux/store";
import { reducer } from "@/redux/reducer";

describe("Cell", () => {
  let store: any;
  beforeEach(() => {
    store = configureStore({
      reducer,
      preloadedState,
      devTools: true,
    });
  });
  afterEach(() => {
    store = null;
  });
  it("empty cell test", () => {
    expect(
      renderer
        .create(
          <Provider store={store}>
            <Cell
              animationSpeed={500}
              isAlive={false}
              x={0}
              y={0}
              idx={0}
              dispatch={jest.fn()}
            />
          </Provider>
        )
        .toJSON()
    ).toMatchSnapshot();
  });
  it("filled cell test", () => {
    expect(
      renderer
        .create(
          <Provider store={store}>
            <Cell
              animationSpeed={500}
              isAlive={true}
              x={0}
              y={0}
              idx={0}
              dispatch={jest.fn()}
            />
          </Provider>
        )
        .toJSON()
    ).toMatchSnapshot();
  });
});
