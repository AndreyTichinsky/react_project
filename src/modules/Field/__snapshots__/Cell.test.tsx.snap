// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cell empty cell test 1`] = `
.emotion-0 {
  border: 1px solid #000;
  box-sizing: border-box;
  width: 10px;
  height: 10px;
  -webkit-transition: background-color 500ms ease;
  transition: background-color 500ms ease;
  background-color: #ffffff;
}

<div
  className="cell cell_0_0 emotion-0"
  onClick={[Function]}
  x={0}
  y={0}
/>
`;

exports[`Cell filled cell test 1`] = `
.emotion-0 {
  border: 1px solid #000;
  box-sizing: border-box;
  width: 10px;
  height: 10px;
  -webkit-transition: background-color 500ms ease;
  transition: background-color 500ms ease;
  background-color: #ffffff;
}

<div
  className="cell cell_0_0 emotion-0"
  onClick={[Function]}
  x={0}
  y={0}
/>
`;
